Save Confirmation - Admin (Part 1 of 2)

msarit opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Part 1 of 2 implementing the [Admin-Save-Confirmation RFC](https://github.com/forem/rfcs/blob/237480c4b915a64fc17df8f525b0a429185a1605/text/0078-admin-save-confirmation.md).
In this PR, I refactor all confirmation steps in the `/admin/customization/config` page to use async calls. I also update all associated `requests` and end-to-end tests.

### NOTES
- Do not let the number of `Files changed` alarm you 😅  For about 10 files, I merely added the `Update Settings` button, and added the `updateConfigurationSettings` callback to the form therein.
- It has been observed that there is no visual indication if an update failed or didn't "take". I have noted this and will address this UX-improvement in a future PR. 👍🏾 

---


## QA Instructions, Screenshots, Recordings

1. Login to `http://localhost:xxxx/admin/customization/config` as a `super_admin`
2. Make various changes to the settings on this page. After clicking the `Update Settings` button each time, there should be NO page refresh, and a snackbar with message `Successfully updated settings` should appear in the bottom-left corner.
3. Run `spec/requests/admin/configs_spec.rb` and all tests under `cypress/integration/seededFlows/adminFlows/config` and make sure they pass.

### UI accessibility concerns?
None. No major UI changes made; most changes are backend

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [X] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [X] I will share this change internally with the appropriate teams
- [ ] I will update the Admin Guide
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## What gif best describes this PR or how it makes you feel?

![Captain Marvel](https://media.giphy.com/media/ZvkFmclQO1ImmRNm0K/giphy-downsized.gif?cid=ecf05e477ld684dzxlok1zvg25q9qkpncuffcqkzf6ygdvrp&rid=giphy-downsized.gif&ct=g)

citizen428 commented on app/controllers/admin/settings/campaigns_controller.rb:
```diff
@@ -1,6 +1,11 @@
     class CampaignsController < Admin::Settings::BaseController
+      private
+
+      def authorization_resource
```
Was there a problem without this? In general, if the entity we deal with matches the controller name, the authorization resource will be correctly inferred, so we only need to add it for controllers where the controller name doesn't match.

aitchiss commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,40 @@ export default class ConfigController extends Controller {
+    } catch (err) {
+      this.displaySnackbar(err);
+    }
+    event.target.querySelector('input[type=submit]').disabled = false;
```
I think we need to set this `disabled` attribute to `true` at the beginning of this function - definitely seems to be possible to hammer the button and trigger a lot of the same network request 😅 

I also wonder (not really for this PR, but just a thought for future iterations) if we want to show some kind of visual indicator to a user that their update is in progress? The process is almost too smooth, so it feels like nothing is happening after I pressed submit

aitchiss replied:
aaaah actually ignore that first part as I just saw your comment about this functionality.

I do think it's a bit confusing at the moment - because I can fire of 4, 5, however many updates at once, and I don't have any indication as to which have succeeded and which have failed

msarit replied:
yeah I'm not sure how to counteract this ... I felt the same way as you 😓

aitchiss replied:
I'm not sure either 🤔 

Maybe the confirmation message could have the input name in it or something e.g. 'Successfully updated suggested users.'?  Maybe there's a way to highlight any inputs we fail to update by adding an `error` class or something 🤔 🤔  I'm not sure! I don't think it necessarily has to be solved in this PR though 🙂

msarit replied:
Ok, I will make a note of this for the near future.

nickytonline replied:
Is disabling the button from an accessibility standpoint OK in this scenario? There are more accessible patterns to handle this or is this OK in this scenario @aitchiss?

aitchiss replied:
It's a good point - the problem with disabling buttons/inputs is it effectively removes them completely from the accessibility tree (until re-enabled). It can be confusing for assistive tech users, but in the case where it happens _after_ submit it's not so bad really (since they already found the button to activate it, it's less confusing to understand why it's now disappearing, if you see what I mean).

I notice that disabling the button on submit is a default Rails behaviour, and we're having to work around it by re-enabling. I'm not sure if it's appropriate for this PR (the submit button is a shared component and maybe in some places we still want the disabled behaviour?) but we could stop Rails from disabling on submit by adding

```erb
<%= f.submit "Update Settings", class: "crayons-btn mt-4", data: {disable_with: false} %>
```

msarit replied:
@aitchiss Yes the Submit button is a shared component, BUT I'm only using it for the Light Confirmation flows (ie this PR 😅 )

So I think this solution you suggest will work just fine. I like that I'm not effecting the button re-enable from Stimulus.

aitchiss commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,40 @@ export default class ConfigController extends Controller {
+
+      const outcome = await response.json();
+
+      if (response.ok) {
+        this.displaySnackbar(outcome.message);
+      } else {
+        this.displaySnackbar(outcome.error);
+      }
```
```suggestion
      this.displaySnackbar(response.ok ? outcome.message : outcome.error);
```

feel free to ignore - I just love adding ternaries for these situations 😄  There's no benefit other than what you prefer looking at though

aitchiss commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,40 @@ export default class ConfigController extends Controller {
+    event.preventDefault();
+    try {
+      const data = new FormData(event.target);
+      const response = await fetch(event.target.action, {
+        method: 'POST',
+        headers: {
+          Accept: 'application/json',
+          'X-CSRF-Token': document.querySelector("meta[name='csrf-token']")
+            ?.content,
+        },
+        body: data,
+        credentials: 'same-origin',
+      });
```
suggestion (not blocking) - you could use the `request` utility function from `app/javascript/utilities/http/request.js`, which adds a lot of these headers & the csrf token for you 🙂

msarit replied:
Ahhh ok, how do I do that? Is there an example in the codebase?

nickytonline replied:
Refactor (blocking): Plus one on the recommendation to use `request`. Using `request` encapsulates the required headers etc.

nickytonline replied:
Search the codebase for ` request(` (space before request), and you'll find a bunch of examples.

aitchiss commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,40 @@ export default class ConfigController extends Controller {
+        this.displaySnackbar(outcome.error);
+      }
+    } catch (err) {
+      this.displaySnackbar(err);
```
I tried to test out the error path by blocking one of the network requests in dev tools - I can see that we're hitting this code with "TypeError: Failed to fetch", but the snackbar shows up as empty 🙃  Not sure what's up here -

https://user-images.githubusercontent.com/20773163/130247152-1e51c6b9-f466-48b0-b03f-8dd5671683f7.mp4

I'm also wondering from the UX side of things, should we be drawing a user's attention to the input field when something doesn't update? Or change the input value back to the old one in the UI to reflect that the request failed? I think it's quite easy to lose track of what failed and where, especially now it's async - if I hit submit and scroll away I have to peep this error message and remember what on earth I was trying to update 😅

msarit replied:
Hmmmm I'm not sure what's up with the `err` not showing in the snackbar 🤔 

I think your UX observations are spot on; I'm not sure how I would "revert" the user's entries back to the old ones without a refresh though...

aitchiss replied:
Yeah it's true, it would add quite a layer of complexity to try to revert the entries (and maybe not worth it anyway - would a user notice it'd changed back?)

I suggested it above but maybe an `error` class could be added to highlight it in red with an `!` icon or something 🤔  Can definitely be a future enhancement though

msarit replied:
Ok I have made a note to address highlighting failed updates. I also fixed the `err` not showing in the snackbar - I needed to call `err.message` 😅

aitchiss commented on app/views/admin/settings/forms/_meta_keywords.html.erb:
```diff
@@ -1,4 +1,6 @@
-<%= form_for(Settings::General.new, url: admin_settings_general_settings_path) do |f| %>
+<%= form_for(Settings::General.new,
+             url: admin_settings_general_settings_path,
+             html: { data: { action: "submit->config#activateLightConfirmation" } }) do |f| %>
```
I love how straightforward this sync up for the light confirm is - so reusable and clear 😭

msarit commented on app/controllers/admin/settings/base_controller.rb:
```diff
@@ -24,6 +40,8 @@ def extra_authorization_and_confirmation
       end
 
       def confirmation_text_valid?
+        return true unless params.key?(:confirmation)
```
The RFC calls for NOT using the Confirmation Text field in Config updates, so I added this guard. We still want the `confirmation_text_valid?` check in place for use within the rest of the Admin portal.

msarit commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,40 @@ export default class ConfigController extends Controller {
+    } catch (err) {
+      this.displaySnackbar(err);
+    }
+    event.target.querySelector('input[type=submit]').disabled = false;
```
For some reason, after `activateLightConfirmation` has fired, the `Update Settings` button remains disabled until after page-refresh. I added this line to re-enable it, so that if the user makes another change before refreshing the page, they can click the button again.

I welcome any solutions that are cleaner than this 🙏🏾

aitchiss commented on cypress/integration/seededFlows/adminFlows/config/authenticationSection.spec.js:
```diff
@@ -31,11 +31,6 @@ describe('Authentication Section', () => {
             .should('not.be.checked')
             .check();
 
-          cy.get('@authSectionForm')
```
I wonder if we should add some new tests for the error paths? We can use Cypress to fake error responses from the server -  e.g. we use intercepts to check some error paths in comment creation

https://github.com/forem/forem/blob/b58fdd013b6e362fe5e120decb5ab9f0cec00375/cypress/integration/seededFlows/articleFlows/commentOnArticle.spec.js#L468-L470

what do you think? Maybe not needed for _every_ input 😅

msarit replied:
I think it's a good idea. I will add some.

aitchiss reviewed (commented):
I gave this a quick spin and it looks to be working really well! I just had a question around the error path - not sure why I'm not seeing an error message 🤔

msarit commented on app/views/admin/settings/forms/_mandatory.html.erb:
```diff
@@ -40,18 +40,8 @@
-      </div>
-    <% end %>
   </fieldset>
+  <% if current_user.has_role?(:super_admin) %>
```
The other partials in the `/admin/customizations/config` view do not have this `<% if current_user.has_role?(:super_admin) %>` conditional in place, but the conditional was present in this partial, so I left it in. @citizen428 advise? I think you last touched this file before I did.

rhymes replied:
As far as I know, all settings inheriting to `Admin::Settings::BaseController`  will go through this authorization check: 

https://github.com/forem/forem/blob/6d2cdeaf41533ff5289e4738783bc7b0e156d60b/app/controllers/admin/settings/base_controller.rb#L22

so I think the if in the partial is redundant, as they won't be able to update anyway. 

I don't think we should hide a submit button, sounds like a dark pattern

nickytonline reviewed (changes requested):
Did a quick pass and will pull this locally to dig deeper. Looking forward to seeing this in production! 😎

msarit commented on spec/requests/admin/configs_spec.rb:
```diff
@@ -807,15 +807,17 @@
-          expect(Settings::SMTP.port).to eq(25)
+
+          expect(Settings::SMTP.address).to eq(default_address)
+          expect(Settings::SMTP.port).to eq(default_port)
```
@citizen428 @rhymes
This expectation on `Settings::SMTP.port` keeps failing CI (although it passes consistently on local):

```
 1) /admin/customization/config POST /admin/customization/config when admin has full permissions including super SMTP configs unsets appropriate SMTP config, and apply default value if applicable
     Failure/Error: expect(Settings::SMTP.port).to eq(default_port)
       expected: 0
            got: 25
```

On local, if I say `expect(Settings::SMTP.port).to eq(25)` (as the CI test seems to want), it fails locally. This is why I set the expectation to expect `default_port`. But this approach fails CI.

How do I make this spec more robust to pass locally and on Travis? 😢  😓  Thank you!

rhymes replied:
So, `Settings::SMTP.port` defaults to `ApplicationConfig["SMTP_PORT"].presence || 25` so it's always 25 unless set to a different value. The reason why you're getting 0 here is because the test does `.to_i` which becomes 0 which is `trueish` in Ruby.  So `0 || 25` will yield 0.

```ruby
default_port = (ApplicationConfig["SMTP_PORT"].presence || 25).to_i
```

is probably what you're looking for.

BTW `ApplicationConfig["SMTP_ADDRESS"].presence || nil` is redundant as `.presence` will return nil if it's not present

Hopefully this help!

msarit commented:
> Refactor (blocking): Plus one on the recommendation to use request. Using request encapsulates the required headers etc.

@nickytonline I'm getting an error: `ReferenceError: getCsrfToken is not defined`. In `request.js` I see this line: `csrfToken = await getCsrfToken()`.

@aitchiss and I read through the codebase and realized that the `request` utility is not available with the `admin` space. Here are her notes on the matter:

```
yeah, it looks like we don't have other uses of request in admin, and it's likely for that reason. I think a good refactor to do (at some point) would be to bring that getCsrfToken function into utils like the request function, but it would have much broader reach than this piece of work
I know Nick mentioned it as blocking but I think he won't have realized this either - personally I think we should do it as a later refactor, and convert other uses of fetch in admin to use request as part of that work
```

Accordingly, I am going to keep my original implementation 😄

nickytonline commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,40 @@ export default class ConfigController extends Controller {
+  async activateLightConfirmation(event) {
+    event.preventDefault();
+    try {
+      const data = new FormData(event.target);
```
nitpick (non-blocking): You can call this `body` and use shorthand notation below, e.g.

```javascript
...
body,
...
```

instead of

```javascript
...
body: data,
...
```

msarit commented:
Okay! All CI passing! 🎉 🎉 🎉 

Tagging @nickytonline and @aitchiss for a final review. Since I'll be OOO starting tomorrow, feel free to merge when the PR is ready. 🙏🏾 

As a way to [draw the user's attention](https://github.com/forem/forem/pull/14418#discussion_r692984340) to settings updates that failed, I had added a `Dismiss` button to the snackbars generated after updates. My expectation was that the snackbar wouldn't fade until the user clicked 'Dismiss', thereby drawing their attention as desired. However, [on Slack we discussed](https://forem-team.slack.com/archives/C7GE84DEJ/p1629734364228400) how the snackbar lifetime is globally coded and cannot be specified locally. So I decided to remove the `Dismiss` buttons. Just FYI - particularly for Suzanne. 👍🏾

aitchiss commented on cypress/integration/seededFlows/adminFlows/config/authenticationSection.spec.js:
```diff
@@ -31,22 +31,22 @@ describe('Authentication Section', () => {
             .click();
 
+          cy.findByTestId('snackbar').within(() => {
+            cy.findByRole('alert').should(
+              'have.text',
+              'Successfully updated settings.',
+            );
```
nice 🔥

aitchiss commented on cypress/integration/seededFlows/adminFlows/config/authenticationSection.spec.js:
```diff
@@ -127,5 +115,28 @@ describe('Authentication Section', () => {
+        cy.findByTestId('authSectionForm').as('authSectionForm');
+
+        cy.get('@authSectionForm').findByText('Authentication').click();
+        cy.get('#facebook-auth-btn').click();
+        cy.get('#settings_authentication_facebook_key').type('randomkey');
+        cy.get('#settings_authentication_facebook_secret').type('randomsecret');
+
+        cy.get('@authSectionForm').findByText('Update Settings').click();
```
just a suggestion and definitely not blocking, but it's a bit more robust to use `cy.findByRole` or `cy.findByLabelText` wherever possible, rather than getting by ID selector or inner text

msarit replied:
Yes @aitchiss Nick also pointed this out to me yesterday 😊 

These cypress tests were existing before my modifications, and the use of `cy.get(#id)` is quite plentiful throughout several files. We decided that making all these tests more robust was beyond the scope of this PR.

aitchiss reviewed (approved):
Took this for a spin again locally and looks good to me - nice one @msarit ✨

msarit commented on spec/requests/admin/configs_spec.rb:
```diff
@@ -807,15 +807,17 @@
         it "unsets appropriate SMTP config, and apply default value if applicable" do
-          Settings::SMTP.address = "smtp.example.com"
-          Settings::SMTP.port = 12_345
+          default_address = ApplicationConfig["SMTP_ADDRESS"].presence || nil
```
Hey MK, locally and on Travis, this spec kept failing, specifically the expectation that the default port was `25` (locally it was `2525` for me). I thought it more robust to expect whatever the default port was on the system running the spec. Same for the default address.

citizen428 replied:
😄 I understand and agree with the reasoning. I was more wondering why we use `ApplicationConfig` directly instead of

```
Settings::SMTP.get_field(:smtp_port)[:default]
```

Which is a bit more general.

msarit replied:
I will change it then @citizen428 😅  I understand the point you are making now.

msarit replied:
Hey @citizen428,

So locally, this is what I get for each of these Settings values:

```
Settings::SMTP.get_field(:smtp_address)[:default]
=> nil

Settings::SMTP.get_field(:smtp_port)[:default]
=> nil

ApplicationConfig["SMTP_PORT"]
=> "2525"

ApplicationConfig["SMTP_ADDRESS"]
=> "smtp.mailtrap.io"
```

The expectations in the relevant `configs_spec.rb` specs are now failing when I use the more general `Settings::SMTP.get_field`.

I'm not sure why these values differ from each other. I think I'll keep use of `ApplicationConfig`; thoughts?

citizen428 replied:
It's fine to use `ApplicationConfig` of course 😃

nickytonline commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,39 @@ export default class ConfigController extends Controller {
+    event.preventDefault();
+    try {
+      const body = new FormData(event.target);
+      const response = await fetch(event.target.action, {
```
I had proposed using our `request utility` class, but it requires a refactor to be able to use it in the admin area, so for this PR, let's keep it as `fetch` for now.

nickytonline commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,39 @@ export default class ConfigController extends Controller {
+
+      const outcome = await response.json();
+
+      outcome.message == null
+        ? this.displaySnackbar(outcome.error)
+        : this.displaySnackbar(outcome.message);
```
Refactor (non-blocking): Use the [nullish-coalescine operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator).

```suggestion
      this.displaySnackbar(outcome.error ?? outcome.message)
```

citizen428 replied:
@nickytonline Isn't your suggestion reversed? The first one says display `.error` when `.message == null`, yours is the other way around, isn't it?

nickytonline replied:
Updated it. 🙃

msarit replied:
Today I learned! 😅  Thanks Nick.

nickytonline commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -93,6 +93,39 @@ export default class ConfigController extends Controller {
+    );
+  }
+
+  async activateLightConfirmation(event) {
```
nitpick (potentially-blocking): It's not clear to me what `activateLightConfirmation` means. Is this method called `activateLightConfirmation` because it's not a form post with a full server-side roundtrip? From what it looks like in the context of where this is called, `updateConfigurationSettings` or `updateSettings` seems more appropriate.

msarit replied:
I will change

rhymes commented on app/controllers/admin/settings/base_controller.rb:
```diff
@@ -22,11 +22,11 @@ def create
 
             format.json do
-              render json: { error: "😭 #{result.errors.to_sentence}" }, status: :unauthorized
+              render json: { error: "❗️❗️ #{result.errors.to_sentence}" }, status: :unauthorized
```
I wonder how these symbols are going to be read by the screen reader. I understand they're added to catch the attention of the user since we don't have a way to differentiate between errors and non errors in Snackbar and we should have a global notification area for errors or something and that's outside the scope wrt the conversation with Pawel. 

I think the Admin team should figure this one out in a future RFC/PR

cc. @aitchiss @ludwiczakpawel

nickytonline replied:
We probably don't need any emojis in the error message.

aitchiss replied:
I would love it if instead of emojis in the text we could have an optional svg/img icon in the snackbar to convey an error message vs a success/info message. You're right @rhymes  in that screen reader behaviour with emojis can vary widely, and what it announces might not be something we intend (e.g. VO on Chrome announces 😭  as "loudly crying face", but this can be different when using a different screen reader/browser/etc)

aitchiss replied:
[Bonus context on how to mark-up emoji accessibly](https://dev.to/s_aitchison/accessible-images-icons-and-emojis-53g7#emojis) 😄

msarit replied:
I think I agree with Nick here - removing the emojis from the error messaging. They were already in place when I started this task; I changed the crying-face emoji to the exclamation points as I thought that was more indicative of an error. However, using the emojis re not worth the a11y drawbacks.

Eventually, I would love our snackbars to be customizable in terms of color perhaps, to indicate a successful transaction, as opposed to a failed one. But that's another PR.

msarit commented:
> This means that one can theoretically skip the config check for everything by removing :confirmation from the request params with a tool like Charles Proxy or Burp Suite. Not the biggest issue but still something to keep in mind.

@citizen428 In response to your comment above MK, I determined that we only use the [confirmation form](https://github.com/forem/forem/blob/main/app/views/admin/settings/_form_submission.html.erb) within the sections at `admin/customization/config`. I checked with Pod AdminExp, and since I'm making all config updates async, I have removed the confirmation form code and refactored accordingly.

citizen428 commented on app/controllers/admin/settings/base_controller.rb:
```diff
@@ -1,37 +1,21 @@
+          render json: { message: "Successfully updated settings." }, status: :ok
         else
-          redirect_to admin_config_path, alert: "😭 #{result.errors.to_sentence}"
+          render json: { error: result.errors.to_sentence }, status: :unauthorized
```
I'm don't think `:unauthorized` is the right status here. You call `authorize_super_admin` before we get here, so at this point, the update failed for other reasons. Usually we use `:unprocessable_entity` I think.

msarit replied:
Yes I hadn’t quite gotten to changing it 😅

citizen428 commented on app/controllers/admin/settings/general_settings_controller.rb:
```diff
@@ -12,9 +12,9 @@ def create
+          render json: { message: "Successfully updated settings." }, status: :ok
         else
-          redirect_to admin_config_path, alert: "😭 #{result.errors.to_sentence}"
+          render json: { error: result.errors.to_sentence }, status: :unauthorized
```
Same re `:unauthorized`.

msarit commented:
@nickytonline Ready for your re-review!

nickytonline reviewed (approved):
Thanks for making the changes @msarit! Looking forward to seeing this in production! 🚀
